### PR TITLE
BUG: Improve thead safety of ObjectFactoryBase::Initialize

### PR DESCRIPTION
**Pull request by blowekamp**
Block concurent threads until initializaiton is completed.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

File: Modules/Core/Common/src/itkObjectFactoryBase.cxx [hunk 1]
```diff
@@ -237,14 +238,19 @@ ObjectFactoryBase::Initialize()
...
 
-  // Atomically set m_Initialized to true. If it was false before, enter the if.
-  if (!m_PimplGlobals->m_Initialized.exchange(true))
+  if (!m_PimplGlobals->m_Initialized)
   {
-    ObjectFactoryBase::InitializeFactoryList();
-    ObjectFactoryBase::RegisterInternal();
+    // only one thread initializing at a time, wait until initialization is completed.
+    std::lock_guard lockGuard(m_PimplGlobals->m_Lock);
+    if (!m_PimplGlobals->m_Initialized)
```
Comment (N-Dekker) [inline Modules/Core/Common/src/itkObjectFactoryBase.cxx:245]: Looks like you're doing _double-checked locking_, which is considered an _anti-pattern_ nowadays! If one thread sets `m_PimplGlobals->m_Initialized = true` while at the same time, another thread tries to evaluate `m_PimplGlobals->m_Initialized` outside the locked region, we have undefined behavior.

File: Modules/Core/Common/src/itkObjectFactoryBase.cxx [hunk 2]
```diff
@@ -237,14 +238,19 @@ ObjectFactoryBase::Initialize()
...
+    {
+      ObjectFactoryBase::InitializeFactoryList();
```
Comment (N-Dekker) [inline Modules/Core/Common/src/itkObjectFactoryBase.cxx:247]: Please consider removing the `ObjectFactoryBase::InitializeFactoryList()` call, as in:
- #4225

File: Modules/Core/Common/src/itkObjectFactoryBase.cxx (hunk 1 above)
Comment (blowekamp) [reply Modules/Core/Common/src/itkObjectFactoryBase.cxx:245]: What is the updated pattern? To ensure it's only initialized once and block other threads until it it initialed?

Comment (dzenanz) [reply Modules/Core/Common/src/itkObjectFactoryBase.cxx:245]: std::call_once?

Comment (blowekamp) [reply Modules/Core/Common/src/itkObjectFactoryBase.cxx:245]: My bigger concern now is if using `std::mutex` during static initialization is OK aka ( defined behavior ).

> evaluate m_PimplGlobals->m_Initialized outside the locked region, we have undefined behavior.

Not sure how undefined the undefined read behavior would be, maybe just not 100% synced between threads. It's will either be true or false, either will be OK.

Comment (N-Dekker) [reply Modules/Core/Common/src/itkObjectFactoryBase.cxx:245]: > std::call_once?

Thanks, @dzenanz. I just gave it a try, using std::call_once. Please check! (Still draft!)
- #4227

**Comment by blowekamp** [comment]
4227 looks like a better approach.

**Pull request closed without merging**